### ISSUE TITLE
destructuring inconsistency with anonymous functions

### ISSUE THREAD

**Issue by non-Jedi** [JuliaLang/julia#36595]
```julia
function f((a, b)) end

g = function ((a, b)) end
```

In the above code, `f` is a function of one argument while `g` is a function of two arguments. I assumed this was intended behavior but Stefan on slack thought it seemed buggy.

Somewhat related to #36581 where @tkf [argues that this behavior should be expected](https://github.com/JuliaLang/julia/issues/36581#issuecomment-655833103).

**Comment by JeffBezanson** [comment]
Yeah, I see the issue, but as tkf pointed out it's due to the difference between function calls and parentheses. `f(x)` adds a call expression around `x`, but `((x))` is the same as `(x)` and `x`.

`function ((a, b))` could potentially be changed, since the `function` keyword could trigger different parsing of parentheses, but changing `((a, b))->...` is harder. Overall I'm not so in favor of changing this.

**Comment by non-Jedi** [comment]
Why is `function((a, b))` (without the space after `function`) parsed the same way? Is that not a call expression despite looking like it?

**Comment by JeffBezanson** [comment]
It's not, since `function` is a reserved word, which makes it special --- more like punctuation than an identifier.

**Comment by tkf** [comment]
> but changing `((a, b))->...` is harder

Is it because all the parsing of parentheses have to wait until the next token is `->` or not?

I was thinking that at least showing warning/error from `((a, b))->...` might be better.  But maybe it's as hard as parsing this as `((a, b),)->...`?

**Comment by lojic** [comment]
Just got bit by this. I'm glad `((p1,p2),) -> ispos(p1) && ispos(p2)` (using the trailing comma) is a workaround, but the inconsistency between named and anonymous functions feels pretty bad.

**Comment by simeonschaub** [comment]
I agree that `((a, b),) -> ...` is a bit ugly and it would be consistent with other function definitions like `f((a, b)) = ...` or `... do (a, b)` to make `((a, b)) -> ...` define a one arg function with destructuring.

> >    but changing ((a, b))->... is harder
> 
> Is it because all the parsing of parentheses have to wait until the next token is -> or not?

This would have to be special-cased, but it shouldn't be too hard to implement if we decide to do this.

One interesting edge case though are anonymous closures using `where` for free type parameters:
```julia
julia> x::T where T -> T
ERROR: syntax: invalid variable expression in "where" around REPL[16]:1
Stacktrace:
 [1] top-level scope at REPL[16]:1

julia> (x::T where T) -> T
#14 (generic function with 1 method)

julia> ans(1)
ERROR: UndefVarError: T not defined
Stacktrace:
 [1] (::var"#14#15")(::Int64) at ./REPL[17]:1
 [2] top-level scope at REPL[18]:1

julia> ((x::T,) where T) -> T
#16 (generic function with 1 method)

julia> ans(1)
Int64

julia> ((x::T,) where T,) -> T
ERROR: syntax: "(x::T,) where T" is not a valid function argument name around REPL[21]:1
Stacktrace:
 [1] top-level scope at REPL[21]:1
```
Not sure what to do about this with that change. Perhaps we could change precedence here to make the first one work, but this might still be breaking some code.

**Comment by orionlee** [comment]
A somewhat related issue. There appears to have some inconsistency (or possibly a separate bug) in handling destructuring between `do` / `for` and anonymous functions. Specifically, the syntax `(a, b)` works for `do` and `for`, but not for anonymous functions. 

```julia
julia> foo, bar = ['a', 'b', 'c'], [1, 2, 3]
(['a', 'b', 'c'], [1, 2, 3])

julia> res = map(zip(foo, bar)) do (f, b)
         (b + 10, f)
       end;

julia> for (f, b) in zip(foo, bar)
         push!(res, (b + 15, f))
       end;

# similar destructuring does not work for anonymous functions.
julia> res = map(((f, b)) -> (b + 20, f), zip(foo, bar));
ERROR: MethodError: no method matching (::var"#11#12")(::Tuple{Char, Int64})
Closest candidates are:
  (::var"#11#12")(::Any, ::Any) at REPL[7]:1

# for anonymous functions, it only accepts single tuple as the argument:
julia> res = map( f_b -> (f_b[2] + 20, f_b[1]), zip(foo, bar));
...
```
The above behavior is found in Julia v1.7.2.
